assayist/processor/base.py

```diff
@@ -0,0 +1,121 @@
+# SPDX-License-Identifier: GPL-3.0+
+
+from abc import ABC, abstractmethod
+import json
+import neomodel
+import os
+
+from assayist.common.models import content, source
+from assayist.processor.configuration import config
+from assayist.processor.logging import log
+
+
+class Analyzer(ABC):
+    """Base Abstract class that analyzers will inherit from."""
+
+    METADATA_DIR = '/metadata/'
+    MESSAGE_FILE = METADATA_DIR + 'message.json'
+    BUILD_FILE = METADATA_DIR + 'buildinfo.json'
+    TASK_FILE = METADATA_DIR + 'taskinfo.json'
+    RPM_FILE = METADATA_DIR + 'rpms.json'
+    ARCHIVE_FILE = METADATA_DIR + 'archives.json'
+    IMAGE_RPM_FILE = METADATA_DIR + 'image-rpms.json'
+    BUILDROOT_FILE = METADATA_DIR + 'buildroot-components.json'
+
+    def main(self):
+        """Call this to run the analyzer."""
+        neomodel.config.DATABASE_URL = config.DATABASE_URL
+        neomodel.config.AUTO_INSTALL_LABELS = True
+        # run the analyzer in a transaction
+        neomodel.db.begin()
+        try:
+            self.run()
+            log.debug("Analyzer completed successfully, committing.")
+            neomodel.db.commit()
+        except Exception as e:
+            log.exception("Error encountered executing Analyzer, rolling back transaction.")
+            neomodel.db.rollback()
+            raise
+
+    @abstractmethod
+    def run(self):
+        """Implement anlyzer code here in your subclass."""
+
+    def read_metadata_file(self, FILE):
+        """Read and return the specified json metadata file or an empty dict."""
+        if os.path.isfile(FILE):
+            with open(FILE, 'r') as f:
+                return json.load(f)
+        else:
+            log.debug("File not found: %s, returning empty dict", FILE)
+            return {}
+
+    def get_or_create_build(self, build_id, build_type):
+        """Get or create a build object."""
+        build = content.Build.nodes.get_or_none(id_=build_id)
+        if build:
+            return build
+
+        return content.Build(id_=build_id, type=build_type).save()
+
+    def get_or_create_rpm_artifact(self, id, name, epoch, version, release, arch, checksum):
+        """Fetch or create an Artifact for this rpm."""
+        # treat empty epochs as zero for consistency
+        epoch = epoch if epoch else '0'
+        nevr = f"{name}-{epoch}:{version}-{release}"
```

Comment:
This looks more like an identifier you're constructing here. Do we have a need for a separate identifier field, which might be distinct from the filename we detect? 

It seems like it might be good for consistency with other tools to maintain whatever conventions we have in the company for fields like `filename` if we can. I was literally thinking this would be something like `maven-3.5.9.noarch-1.rpm`. I'd like to know if that's not what we need, though.